### PR TITLE
Add build2.d and reference buildlib

### PR DESCRIPTION
**Pull request by marler8997**
Created a new repository http://github.com/dragon-lang/buildlib and put common build code in there.  The new `build2.d` script references that code.

### CODE REVIEW

File: buildlib/build/log.d [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    if (verboseEnabled)
+        writefln(args);
+}
+
+/**
+Logs the message to stdout.
+Params:
+    args = the data to write to the log
+*/
+void logf(T...)(T args)
```
Comment (wilzbach) [inline buildlib/build/log.d]: What's the point of this? And why not `alias ...`?

File: buildlib/build/proc.d [hunk 2]
```diff
@@ -0,0 +1,9 @@
+module build.proc;
+
+auto execute(T...)(scope const(char[])[] args, T extra)
+{
+    import std.proc;
```
Comment (wilzbach) [inline buildlib/build/proc.d]:
```suggestion
    import std.process : execute;
```

File: buildlib/build/string.d [hunk 3]
```diff
@@ -0,0 +1,25 @@
+module build.string;
+
+auto firstLine(T)(T str)
+{
+    import std.string;
```
Comment (wilzbach) [inline buildlib/build/string.d]:
```suggestion
    import std.string : indexOf, stripRight;
```

File: buildlib/build/string.d [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    auto newlineIndex = str.indexOf("\n");
+    if (newlineIndex < 0)
+        return str;
```
Comment (wilzbach) [inline buildlib/build/string.d]: Why not `stripLeft`?

File: buildlib/build/string.d [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    return str[0 .. newlineIndex].stripRight("\r");
+}
+
+auto sliceInsideQuotes(inout(char)[] str)
+{
+    import std.format;
+
+    auto firstQuote = str.indexOf('"');
+    if (firstQuote < 0)
+        throw new Exception(format("string did not contain quotes '%s'", str));
```
Comment (wilzbach) [inline buildlib/build/string.d]: Why not use `enforce`?

File: buildlib/build/string.d [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    auto result = str[firstQuote + 1 .. $];
+    auto secondQuote = result.indexOf('"');
+    if (secondQuote < 0)
+        throw new Exception(format("string did not end with quote '%s'", str));
```
Comment (wilzbach) [inline buildlib/build/string.d]: Why not use `enforce`?

File: buildlib/build/string.d [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    auto newlineIndex = str.indexOf("\n");
+    if (newlineIndex < 0)
+        return str;
+    return str[0 .. newlineIndex].stripRight("\r");
+}
+
+auto sliceInsideQuotes(inout(char)[] str)
+{
+    import std.format;
```
Comment (wilzbach) [inline buildlib/build/string.d]:
```suggestion
    import std.format : format;
```

File: buildlib/build/log.d (hunk 1 above)
Comment (marler8997) [reply buildlib/build/log.d]: Allows you to edit all log statements in one place.  I might add a prefix to all log statements for example, or direct output to a different file.

File: buildlib/build/string.d (hunk 4 above)
Comment (wilzbach) [reply buildlib/build/string.d]: Ah I was referring to stripLeft's second argument (you can do, e.g. `str.stripLeft("\n"))`), but in this case it looks like you want `until`.

File: src/build2.d [hunk 8]
```diff
@@ -0,0 +1,1051 @@
...
+{
+    if (!__traits(getMember, Env, member))
+        return false;
+    if (__traits(getMember, Env, member) != "1")
+        throw fatal("Environment Variable '%s' must either be '1' or unset, but is '%s",
+            member, __traits(getMember, Env, member));
+    return true;
+}
+
+// TODO: already a library function that does this?
```
Comment (wilzbach) [inline src/build2.d]: Memoize

File: src/build2.d [hunk 9]
```diff
@@ -0,0 +1,1051 @@
...
+    if (hostOS == "solaris")
+        return ["isainfo", "-n"].execute.output.archFromUname;
+    else if (hostOS == "windows" || hostOS == "Windows_NT")
+        return ["wmic", "OS", "get", "OSArchitecture"].execute.output.archFromUname;
+    else
+        return ["uname", "-m"].execute.output.archFromUname;
+}
+string determineTargetOS()
+{
+    return Env.TARGET_OS ? Env.TARGET_OS : getHostOS;
```
Comment (wilzbach) [inline src/build2.d]: Why not `||`?

**Comment by marler8997** [comment]
Cleaning old PRs to clean up my pull request list.

**Pull request closed without merging**